service/manager.go

```diff
@@ -0,0 +1,301 @@
+		return err
+	}
+	service.Hash = hex.EncodeToString(hash)
+
+	// replace default env with new one.
+	defenv := xos.EnvSliceToMap(service.Configuration.Env)
+	service.Configuration.Env = xos.EnvMapToSlice(xos.EnvMergeMaps(defenv, env))
+
+	image, err := m.c.Build(contextDir)
+	if err != nil {
+		return err
+	}
+
+	service.Configuration.Image = image
+	return nil
+}
+
+// Start see Manager.Start.
+func (m *ContainerManager) Start(service *Service) error {
+	if err := m.Status(service); err != nil {
+		return err
+	}
+	if service.Status == StatusRunning {
+		return nil
+	}
+	if service.Status == StatusPartial {
+		if err := m.Stop(service); err != nil {
+			return err
+		}
+	}
+
+	networkID, err := m.c.CreateNetwork(service.namespace())
+	if err != nil {
+		return err
+	}
+	sharedNetworkID, err := m.c.SharedNetworkID()
+	if err != nil {
+		return err
+	}
+	_, port, _ := xnet.SplitHostPort(m.cfg.Server.Address)
+	endpoint := m.cfg.Core.Name + ":" + strconv.Itoa(port)
+	start := func(dep *Dependency, namespace []string) error {
+		_, err := m.c.StartService(container.ServiceOptions{
+			Namespace: namespace,
+			Labels: map[string]string{
+				"mesg.service": service.Name,
+				"mesg.hash":    service.Hash,
+				"mesg.sid":     service.Sid,
+				"mesg.core":    m.cfg.Core.Name,
+			},
+			Image:   dep.Image,
+			Args:    dep.Args,
+			Command: dep.Command,
+			Env: xos.EnvMergeSlices(dep.Env, []string{
+				"MESG_TOKEN=" + service.Hash,
+				"MESG_ENDPOINT=" + endpoint,
+				"MESG_ENDPOINT_TCP=" + endpoint,
+			}),
+			Mounts: service.volumes(dep.Key),
+			Ports:  dep.ports(),
+			Networks: []container.Network{
+				{ID: networkID, Alias: dep.Key},
+				{ID: sharedNetworkID},
+			},
+		})
+		return err
+	}
+
+	if err := start(service.Configuration, service.namespace()); err != nil {
+		return err
+	}
+
+	// BUG: https://github.com/mesg-foundation/core/issues/382
+	// After solving this by docker, switch back to deploy in parallel
+	for _, dep := range service.Dependencies {
+		if err := start(dep, depNamespace(service.Hash, dep.Key)); err != nil {
+			return err
+		}
+
+		if err != nil {
+			m.Stop(service)
+			return err
+		}
+	}
+	service.Status = StatusRunning
+	return nil
+}
+
+// Stop see Manager.Stop.
+func (m *ContainerManager) Stop(service *Service) error {
+	var (
+		wg   sync.WaitGroup
+		errs xerrors.SyncErrors
+
+		stop = func(namespace []string) {
+			defer wg.Done()
+			if err := m.c.StopService(namespace); err != nil {
+				errs.Append(err)
+			}
+		}
+	)
+
+	wg.Add(len(service.Dependencies) + 1)
+	for _, dep := range service.Dependencies {
+		go stop(depNamespace(service.Hash, dep.Key))
+	}
+	go stop(service.namespace())
+	wg.Wait()
+
+	if err := errs.ErrorOrNil(); err != nil {
+		return err
+	}
+	if err := m.c.DeleteNetwork(service.namespace()); err != nil {
+		return err
+	}
+	service.Status = StatusStopped
+	return nil
+}
+
+// Delete see Manager.Delete.
+func (m *ContainerManager) Delete(service *Service) error {
```

Comment:
Maybe `DeleteVolumes`. This doesn't delete the service so that's confusing

Reply:
So this is the implementation of the `Manager` interface

```go
  // Delete deletes all persistent data used by service.
  Delete(service *Service) error
```

So for Container, it happens to delete volumes but for other deploy methods, it could be something else.

Reply:
Instead of having the stop and delete logic in API, it could be done at the service manager.
Stop is just stopping the service.
Delete: stop + delete persistent data.

Reply:
No, it's manager part to have deleted and stop function separated. so api (or any other packge) can use them, and there they should be splited.